### ISSUE TITLE
Postgres connector

### ISSUE THREAD

**Issue by jvvasques** [akka/alpakka#67]
I can contribute with a Postgres connector. Is there any interest?

**Comment by johanandren** [comment]
What would it do? A source of query results using the async postgres driver?

**Comment by jvvasques** [comment]
Yes, that was my idea. In terms of design, I'd follow the same structure that you guys have on the Cassandra one

**Comment by ktoso** [comment]
For JDBC actually https://github.com/slick/slick is an existing integration.
We should link to it :)

**Comment by jvvasques** [comment]
Good point @ktoso !

**Comment by jvvasques** [comment]
I'll put together a first version and issue an early PR on the weekend using Slick ;)

**Comment by johanandren** [comment]
For slick there would not need to be a connector as it already implements reactive sockets, so just a link. Possibly some examples but I'm not sure where we'd put those.

**Comment by ktoso** [comment]
^ reactive streams ;)
We could quickly show how to wrap it into an Akka Stream (`fromPublisher`) though

**Comment by matsluni** [comment]
When I looked at it I just found out that there is a `DatabasePublisher` (http://slick.lightbend.com/doc/3.1.1/dbio.html#streaming) is there also a `Sink` (with some kind of back pressure)?

**Comment by ktoso** [comment]
AFAIR there isn't a Sink for inserts in Slick, but I may be wrong - asking for a friend @szeiger ;-)
I think it'd be more universally useful to work on extending Slick with that than adding a one-off integration in Alpakka, WDYT?

**Comment by szeiger** [comment]
This is still missing from Slick and would be a very useful addition. There's no reason to restrict it to inserts though. What we really need is a way to run DBIOActions with back-pressure support, i.e. a `Subscriber[DBIOAction]`. Currently there's only `db.run` which simply fails when the queue is full. This is great for web applications (where you want to fail fast instead of enqueuing even more work when the load is too high) but doesn't work well for batch.

The current recommended practice is to use Akka Streams and wrap the `db.run` call in `mapAsync` but that works only for cases where your batch has full control over a `Database` instance. You cannot share a `Database` between multiple batch operations without strictly splitting up the queue size between them, and you cannot share it at all between a batch operation and non-batch calls.

**Comment by schmitch** [comment]
Actually we needed something like that for CSV generation and we used a *pretty simple* GraphStage to do so, it's quick and dirty but actually it does the job: https://gist.github.com/schmitch/ae48eebf76339896680aa3cb5a70bdec
Guess something like that is super useful. Probably would've been even more useful if slick/quill/others would've have an easy integration. (our use case is not general, since we just need to close the connection, once done).

**Comment by klpx** [comment]
Hi! I already have a postgres jdbc sink connector here: https://github.com/klpx/akka-streams-postgresql-copy and use it in production.
In case of using streams i think COPY is good way for bulk inserts. But my connector doesnt have some stream partitioning for case of long running streams.
I can adapt it for Alpakka code structure and create PR if someone is interested in that.

**Comment by matsluni** [comment]
Hi, there is a PR for Slick [Slick#1675](https://github.com/slick/slick/pull/1675) to add a Database Subscriber. But I don't know if/when it will be merged.

**Comment by patriknw** [comment]
@klpx integration with Postgres is very interesting. Is it using Postgress async APIs?

**Comment by klpx** [comment]
No, it uses JDBC in async manner.
